File: ts-sdk/whirlpool/tests/pool.test.ts
import { describe, it, beforeAll } from "vitest";
import {
  fetchConcentratedLiquidityPool,
  fetchSplashPool,
  fetchWhirlpoolsByTokenPair,
} from "../src/pool";
import {
  getTestContext,
  rpc,
  TEST_WHIRLPOOL_DEPLOYMENTS,
} from "./utils/mockRpc";
import assert from "assert";
import { SPLASH_POOL_TICK_SPACING } from "../src/config";
import type { Address } from "@solana/kit";
import { setupMint } from "./utils/token";
import { setupWhirlpool } from "./utils/program";
import { getWhirlpoolAddress } from "@orca-so/whirlpools-client";

// Ensure both deployments are initialized so the parametrized describe.each below
// sees real config addresses, not the static `mainnet` placeholders.
await getTestContext();

describe.each(TEST_WHIRLPOOL_DEPLOYMENTS)(
  "Fetch Pool ($programId)",
  (whirlpoolDeployment) => {
    let mintA: Address;
    let mintB: Address;
    let defaultPool: Address;
    let concentratedPool: Address;
    let splashPool: Address;

    beforeAll(async () => {
      mintA = await setupMint();
      mintB = await setupMint();
      concentratedPool = await setupWhirlpool(mintA, mintB, 64, {
        whirlpoolDeployment,
      });
      defaultPool = await getWhirlpoolAddress(
        mintA,
        mintB,
        128,
        whirlpoolDeployment,
      ).then((x) => x[0]);
      splashPool = await setupWhirlpool(
        mintA,
        mintB,
        SPLASH_POOL_TICK_SPACING,
        { whirlpoolDeployment },
      );
    });

    it("Should be able to fetch a splash pool", async () => {
      const pool = await fetchSplashPool(
        rpc,
        mintA,
        mintB,
        whirlpoolDeployment,
      );
      assert.strictEqual(pool.initialized, true);
      assert.strictEqual(pool.liquidity, 0n);
      assert.strictEqual(pool.tickSpacing, SPLASH_POOL_TICK_SPACING);
      assert.strictEqual(pool.address, splashPool);
      assert.strictEqual(pool.tokenMintA, mintA);
      assert.strictEqual(pool.tokenMintB, mintB);
      assert.strictEqual(pool.feeRate, 1000);
      assert.strictEqual(pool.protocolFeeRate, 100);
      assert.strictEqual(
        pool.whirlpoolsConfig,
        whirlpoolDeployment.configAddress,
      );
    });

    it("Should be able to fetch a concentrated liquidity pool", async () => {
      const pool = await fetchConcentratedLiquidityPool(
        rpc,
        mintA,
        mintB,
        64,
        whirlpoolDeployment,
      );
      assert.strictEqual(pool.initialized, true);
      assert.strictEqual(pool.liquidity, 0n);
      assert.strictEqual(pool.tickSpacing, 64);
      assert.strictEqual(pool.address, concentratedPool);
      assert.strictEqual(pool.tokenMintA, mintA);
      assert.strictEqual(pool.tokenMintB, mintB);
      assert.strictEqual(pool.feeRate, 300);
      assert.strictEqual(pool.protocolFeeRate, 100);
      assert.strictEqual(
        pool.whirlpoolsConfig,
        whirlpoolDeployment.configAddress,
      );
    });

    it("Should be able to try fetching a non-existent pool", async () => {
      const pool = await fetchConcentratedLiquidityPool(
        rpc,
        mintA,
        mintB,
        128,
        whirlpoolDeployment,
      );
      assert.strictEqual(pool.initialized, false);
      assert.strictEqual(pool.tickSpacing, 128);
      assert.strictEqual(pool.address, defaultPool);
      assert.strictEqual(pool.tokenMintA, mintA);
      assert.strictEqual(pool.tokenMintB, mintB);
      assert.strictEqual(pool.feeRate, 1000);
      assert.strictEqual(pool.protocolFeeRate, 100);
      assert.strictEqual(
        pool.whirlpoolsConfig,
        whirlpoolDeployment.configAddress,
      );
    });

    it("Should be able to fetch all pools for a pair", async () => {
      const pools = await fetchWhirlpoolsByTokenPair(
        rpc,
        mintA,
        mintB,
        whirlpoolDeployment,
      );
      assert.strictEqual(pools.length, 3);

      // Note: we use find because ordering is not guaranteed
      const pool0 = pools.find((p) => p.tickSpacing === 64);
      assert.strictEqual(pool0?.initialized, true);
      assert.strictEqual(pool0?.tickSpacing, 64);

      const pool1 = pools.find(
        (p) => p.tickSpacing === SPLASH_POOL_TICK_SPACING,
      );
      assert.strictEqual(pool1?.initialized, true);
      assert.strictEqual(pool1?.tickSpacing, SPLASH_POOL_TICK_SPACING);

      const pool2 = pools.find((p) => p.tickSpacing === 128);
      assert.strictEqual(pool2?.initialized, false);
      assert.strictEqual(pool2?.tickSpacing, 128);
    });
  },
);

File: rust-sdk/whirlpool/src/tests/pool_test.rs
use super::utils::{
    program::setup_whirlpool,
    rpc::RpcContext,
    token::{setup_ata_with_amount, setup_mint_with_decimals},
};
use crate::{
    fetch_concentrated_liquidity_pool, fetch_splash_pool, fetch_whirlpools_by_token_pair, PoolInfo,
    SPLASH_POOL_TICK_SPACING,
};
use serial_test::serial;

#[tokio::test]
#[serial]
async fn test_fetch_splash_pool() {
    let ctx = RpcContext::new().await;
    let mint_a = setup_mint_with_decimals(&ctx, 9).await.unwrap();
    let mint_b = setup_mint_with_decimals(&ctx, 9).await.unwrap();

    setup_ata_with_amount(&ctx, mint_a, 500_000_000_000)
        .await
        .unwrap();
    setup_ata_with_amount(&ctx, mint_b, 500_000_000_000)
        .await
        .unwrap();

    let splash_pool = setup_whirlpool(&ctx, mint_a, mint_b, SPLASH_POOL_TICK_SPACING)
        .await
        .unwrap();

    if let PoolInfo::Initialized(pool) = fetch_splash_pool(&ctx.rpc, mint_a, mint_b).await.unwrap()
    {
        assert_eq!(pool.data.liquidity, 0);
        assert_eq!(pool.data.tick_spacing, SPLASH_POOL_TICK_SPACING);
        assert_eq!(pool.address, splash_pool);
        assert_eq!(pool.data.token_mint_a, mint_a);
        assert_eq!(pool.data.token_mint_b, mint_b);
        assert_eq!(pool.data.fee_rate, 1000);
        assert_eq!(pool.data.protocol_fee_rate, 0);
    } else {
        panic!("Expected initialized pool");
    }
}

#[tokio::test]
#[serial]
async fn test_fetch_concentrated_liquidity_pool() {
    let ctx = RpcContext::new().await;
    let mint_a = setup_mint_with_decimals(&ctx, 9).await.unwrap();
    let mint_b = setup_mint_with_decimals(&ctx, 9).await.unwrap();

    setup_ata_with_amount(&ctx, mint_a, 500_000_000_000)
        .await
        .unwrap();
    setup_ata_with_amount(&ctx, mint_b, 500_000_000_000)
        .await
        .unwrap();

    let concentrated_pool = setup_whirlpool(&ctx, mint_a, mint_b, 64).await.unwrap();

    if let PoolInfo::Initialized(pool) =
        fetch_concentrated_liquidity_pool(&ctx.rpc, mint_a, mint_b, 64)
            .await
            .unwrap()
    {
        assert_eq!(pool.data.liquidity, 0);
        assert_eq!(pool.data.tick_spacing, 64);
        assert_eq!(pool.address, concentrated_pool);
        assert_eq!(pool.data.token_mint_a, mint_a);
        assert_eq!(pool.data.token_mint_b, mint_b);
        assert_eq!(pool.data.fee_rate, 300);
        assert_eq!(pool.data.protocol_fee_rate, 0);
    } else {
        panic!("Expected initialized pool");
    }
}

#[tokio::test]
#[serial]
async fn test_fetch_non_existent_pool() {
    let ctx = RpcContext::new().await;
    let mint_a = setup_mint_with_decimals(&ctx, 9).await.unwrap();
    let mint_b = setup_mint_with_decimals(&ctx, 9).await.unwrap();

    setup_ata_with_amount(&ctx, mint_a, 500_000_000_000)
        .await
        .unwrap();
    setup_ata_with_amount(&ctx, mint_b, 500_000_000_000)
        .await
        .unwrap();

    if let PoolInfo::Uninitialized(pool) =
        fetch_concentrated_liquidity_pool(&ctx.rpc, mint_a, mint_b, 128)
            .await
            .unwrap()
    {
        assert_eq!(pool.tick_spacing, 128);
        assert_eq!(pool.token_mint_a, mint_a);
        assert_eq!(pool.token_mint_b, mint_b);
        assert_eq!(pool.fee_rate, 1000);
        assert_eq!(pool.protocol_fee_rate, 0);
    } else {
        panic!("Expected uninitialized pool");
    }
}

#[tokio::test]
#[serial]
async fn test_fetch_all_pools_for_pair() {
    let ctx = RpcContext::new().await;
    let mint_a = setup_mint_with_decimals(&ctx, 9).await.unwrap();
    let mint_b = setup_mint_with_decimals(&ctx, 9).await.unwrap();

    setup_ata_with_amount(&ctx, mint_a, 500_000_000_000)
        .await
        .unwrap();
    setup_ata_with_amount(&ctx, mint_b, 500_000_000_000)
        .await
        .unwrap();

    // Create pools with different tick spacings
    let concentrated_pool = setup_whirlpool(&ctx, mint_a, mint_b, 64).await.unwrap();
    let splash_pool = setup_whirlpool(&ctx, mint_a, mint_b, SPLASH_POOL_TICK_SPACING)
        .await
        .unwrap();

    let pools = fetch_whirlpools_by_token_pair(&ctx.rpc, mint_a, mint_b)
        .await
        .unwrap();

    assert_eq!(pools.len(), 3); // 2 initialized + 1 uninitialized (128 tick spacing)

    // Verify concentrated liquidity pool
    let concentrated = pools
        .iter()
        .find(|p| match p {
            PoolInfo::Initialized(p) => p.data.tick_spacing == 64,
            _ => false,
        })
        .unwrap();
    if let PoolInfo::Initialized(pool) = concentrated {
        assert_eq!(pool.address, concentrated_pool);
        assert_eq!(pool.data.fee_rate, 300);
    }

    // Verify splash pool
    let splash = pools
        .iter()
        .find(|p| match p {
            PoolInfo::Initialized(p) => p.data.tick_spacing == SPLASH_POOL_TICK_SPACING,
            _ => false,
        })
        .unwrap();
    if let PoolInfo::Initialized(pool) = splash {
        assert_eq!(pool.address, splash_pool);
        assert_eq!(pool.data.fee_rate, 1000);
    }

    // Verify uninitialized pool
    let uninitialized = pools
        .iter()
        .find(|p| match p {
            PoolInfo::Uninitialized(p) => p.tick_spacing == 128,
            _ => false,
        })
        .unwrap();
    if let PoolInfo::Uninitialized(pool) = uninitialized {
        assert_eq!(pool.fee_rate, 1000);
    }
}
